projects/mesontarget.vala

```diff
@@ -0,0 +1,160 @@
+
+            // get all flags
+            for (int i=0; i<target_source.parameters.length; i++) {
+                string param = target_source.parameters[i];
+                if (/^--\w*[\w-]*\w+$/.match (param)) {
+                    if (param == "--pkg" || param == "--vapidir" ||
+                        param == "--girdir" || param == "--metadatadir" ||
+                        param == "--gresourcesdir" || param == "--define") {
+                        if (i+1 < target_source.parameters.length) {
+                            // the next argument is the value
+                            if (!args.has_key (param))
+                                args[param] = new ArrayList<string> ();
+                            args[param].add (target_source.parameters[i+1]);
+                            i++;
+                        }
+                    } else if (param == "--experimental") {
+                        experimental = true;
+                    } else if (param == "--experimental-non-null") {
+                        experimental_non_null = true;
+                    } else if (param == "--profile") {
+                        if (i+1 < target_source.parameters.length) {
+                            // the next argument is the value
+                            string profile_str = target_source.parameters[i+1];
+                            if (profile_str == "posix")
+                                profile = Vala.Profile.POSIX;
+                            i++;
+                        }
+                    } else if (param == "--abi-stability") {
+                        abi_stability = true;
+                    } else {
+                        debug (@"MesonTarget: ignoring argument `$param'");
+                    }
+                } else {
+                    int idx = param.index_of ("=");
+                    if (idx != -1) {
+                        // --[param_name]={value}
+                        string param_name = param[0:idx];
+                        if (!args.has_key (param_name))
+                            args[param_name] = new ArrayList<string> ();
+                        args[param_name].add (param.substring (idx+1));
+                    } else {
+                        debug (@"MesonTarget: ignoring argument `$param'");
+                    }
+                }
+            }
+
+            var compilation = new Compilation (this, experimental,
+                    experimental_non_null,
+                    profile, abi_stability,
+                    args["--pkg"],
+                    args["--vapidir"],
+                    args["--girdir"],
+                    args["--metadatadir"],
+                    args["--gresourcesdir"],
+                    args["--define"]);
+
+            // add source files to compilation
+            foreach (string source in target_source.sources) {
+                if (!Path.is_absolute (source))
+                    source = Path.build_filename (build_dir, source);
```

Comment:
are the paths ever going to be non-absolute?

Reply:
Yes, sometimes they are.

Reply:
in what cases

Reply:
I don't remember exactly which, and I'm having trouble reproducing them with any of the repos I'm testing VLS with (geary, granite, VLS), but I know I've encountered this before. I don't think there's any downside to having these lines in, since even if they're a waste of time, they're only executed once on server initialization.